### ISSUE TITLE
Add a number or a symbol above/below a candle chart with mplfinance?

### ISSUE THREAD

**Issue by manuelwithbmw** [matplotlib/mplfinance#97]
Hi, is there a way to add on top of a day candle a number ('1', '2', .... , '9', '10', ... ) or other symbols like '+' all this on top of the HIGH or beneath the LOW of the candle? Or is there a way of also drawing on top/below an Arrow  going down/up towards the body candle. 
![Screenshot 2020-04-17 at 00 58 28](https://user-images.githubusercontent.com/63164573/79517638-ac007a80-8046-11ea-9c20-2bbd30f08dda.png)

**Comment by DanielGoldfarb** [comment]
presently there is no way to do this.   **[There is a similar request here](https://github.com/matplotlib/mplfinance/issues/49#issuecomment-606124101)**, and **[this enhancement to allow users to pass in their own Figure and Axes](https://github.com/matplotlib/mplfinance/issues/17)** should provide a good work around as well.  (Expected completion approximately the end of May or beginning of June.

I'm not sure (haven't tried it yet), but you may be able to take advantage of **[this feature](https://github.com/matplotlib/mplfinance/issues/46)** to be able to add the text as well.

Finally, now that I think of it, you should be able to do this **[with the existing `addplot` feature](https://github.com/matplotlib/mplfinance/blob/master/examples/addplot.ipynb).**  Read through that example notebook.  I think if you use a scatter plot, and set the symbols to be the numbers and arrows, you can accomplish what you want, but you may have to call `make_addplot()` several times.

**Comment by manuelwithbmw** [comment]
Hi Daniel, thanks for the prompt answer. I had already built a scatter plot via make_addplot() and having triangles, or squares as marker signals (actually the overall overlapped chart to my candlestick is a signal chart basically). I had tried markers as numbers and been refused, I will play a little more with it. And will look at your other suggested features

**Comment by DanielGoldfarb** [comment]
For using markers, these references may help:

https://matplotlib.org/3.1.0/gallery/lines_bars_and_markers/marker_reference.html

https://matplotlib.org/3.2.1/tutorials/text/mathtext.html  (_note sections on numbers and arrows_)

https://stackoverflow.com/questions/44726675/custom-markers-using-python-matplotlib

https://stackoverflow.com/questions/14324270/matplotlib-custom-marker-symbol

**Comment by manuelwithbmw** [comment]
Oh Daniel, that's a fantastic help, thank you!

**Comment by fxhuhn** [comment]
@manuelwithbmw it would be greate if you share your solution with us. :-)

**Comment by manuelwithbmw** [comment]
So, thanks to above marker reference links that Daniel gave (Thanks!), i managed to draw numbers and arrows on top and below certain candles, with the aim to represent a trading signal of some sort.  I used the scatter method plot (not ideal for signals as all markers are equal per every candle) within a mpf.plot call with an addplot() in it and making use of make_addplot().

- For numbers, I used the following notation to be used within "markers" of a make_addplot:

Valid examples: marker='$1$' or marker='$2$' or ... marker='$9$'

- For arrows, I used the following notations which seems to be valid for "markers" of a make_addplot:

Valid examples: marker=r'$\Uparrow$' or marker=r'$\Downarrow$' or marker=r'$\uparrow$' or marker=r'$\downarrow$'

- I then designed BUY signals in order to be i.e. 5% below the Low of the relevant candle
buy_setup_signal = signal * low * 0.95
- I designed SELL signals in order to be 5% above High of the relevant candle
sell_setup_signal = signal * high * 1.05

As said, to plot all that, I created additional (to candle data frame df) plots for my Sell and Buy signals grouped in a single Dictionary (or whatever it is), like below:

apd = [mpf.make_addplot(sell_signal1, scatter=True,
                        markersize=50, marker='$9$'),
       mpf.make_addplot(buy_signal1, scatter=True,
                        markersize=50, marker='$9$'),
       mpf.make_addplot(buy_signal2, scatter=True,
                        markersize=100, marker=r'$\Uparrow$'),
       mpf.make_addplot(sell_signal2, scatter=True, markersize=50, marker='$5$')]

- I Plotted both stock candle and Buy / Sell signals charts with calls to that Dictionary as addplot:
mpf.plot(df, type='candle', ylabel='Price', ylabel_lower='Volume', mav=(21, 50, 200), volume=True, figscale=0.8, addplot=apd)

I noticed Numbers and Arrows can also be used in charts labels if you like:
Examples: ylabel=r'$\Uparrow$' or ylabel=r'$\Downarrow$' or ylabel=r'$\uparrow$' or ylabel=r'$\downarrow$'

Results shown in the attached pictures
![Arrows and numbers 1](https://user-images.githubusercontent.com/63164573/79642183-fcd1b980-8193-11ea-9f58-8e213eff0802.png)
![Arrows and numbers 2](https://user-images.githubusercontent.com/63164573/79642188-fe9b7d00-8193-11ea-930d-e797341746b9.png)

A scatter not ideal as Ideally I'd treat every candle separately with different numbers, while scatter markers are always the same for the overall chart. And I cannot have 20 subplots I suppose, I will try.

Hope it helps @fxhuhn !

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw - Thank you so much for sharing!  (@fxhuhn Thanks for asking for it)

@manuelwithbmw - Can you please explain "And I cannot have 20 subplots I suppose, I will try."  ... are you talking about "`make_addplot()`" (which are not actually subplots, so 20 should be fine ... tedious, but fine.  they can all be put into a list and pass the list into `mpf.plot()`.

Now, regarding where we might go from here.  I realize calling `mpf.make_addplot()` for each different marker you want is tedious (let alone figuring their locations).  Would you like to propose what you would ideally want the interface to be?   I had originally been thinking regarding #49 to have the user pass in something like `trades=list_of_trades` where list_of_trades would be a list of mostly zeros with positive numbers for buys and negative numbers for sells.  Thoughts?

**Comment by manuelwithbmw** [comment]
Oh yes I articulated improperly, they are not subplots, they are just additional plots to my df.
Basically every number under/above a candle will be a call to make_addplot(), and it represents a reference to a signal. But I need a cluster of those signals, meaning that I need 10-20 candles with numbers above/below before actually having a real actionable signal.
So yes i will put all of them into a list ("apd" in my example above) and will pass this to mpf.plot(). Not particularly tedious in itself, the only prob is I need to calc and handle an appropriate series on NaN (or zeros) plus my number/arrow on/below the correct candle for every one of them, every series. Yes figuring their candle location is the hard part and the fact that the number of elements of this Series has to correspond to len(df) in a different way for every one is driving me mad. Not sure I explained myself.
To be clearer, in my case, the complexity of the indicator is that I need to plot 10-20 numbers/arrow according to certain rules, before actually having a real actionable trading signal, which would be the last of those designed numbers above/below a certain candle.

Yeah, passing in a list_of_trades as a list of mostly zeros (or NaN? Zeros probably better) with positive numbers for buys and negative numbers for sells would work OK.

In my example above signal1 (in "buy_signal = signal1 * low * 0.95) is exactly that, a series of 1s and NaNs like: 

### signal1= [nan nan nan nan nan nan nan nan 1 1 1 nan nan nan 1 1 1 1 1 nan nan nan nan nan nan nan nan nan nan nan nan 1 1 1 1 nan]
where 1s represent the position of a component of my buy trade. And

signal2 (in "sell_signal = signal2 * high * 1.05) is a series of 1s and NaNs like: 

### signal2= [nan 1 1 1 1 1 nan nan 1 1 1 1 nan nan 1 1 1 1 1 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan]
where 1s represent the position of a component of my sell trade.

What I'd like is a function able to plot the numbers/arrows more easily. The real prob for now is that number of all markers have to correspond -obviously- to len(df) and this is hard when you have a math logic embedded to design where the 1s have to be and have not.

So ideally passing in the list_of_trades coupled with different markers for any trade would be the optimal for me, but probably too crazy to program, did I make it overly complicated? Thanks

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw,

It all sounds reasonable to want to do.  My only comment is you seem stuck on the idea that making your signal list the same lengh as the dataframe is somewhat of a challenge.  I would suggest thinking of the signal generation somewhat differently.  

It seems to me (please corrrect me if I am wrong) that to generate each of the signals, you need to examine every row (date) in your dataframe prices (even if for some of those rows you do nothing, or return zero or NaN or None).  If so, you many consider using `Pandas.Dataframe.apply()` to do this, and place the result from the `apply()` call into a new column in the same data frame.  Then every signal is automatically the correct length.
    
If it is not clear what I mean, then if I can see your code I may pehaps be able to show specifically for your use-case what I mean.

**Comment by manuelwithbmw** [comment]
It is exactly as you say. I didn't know about Pandas.Dataframe.apply(), this sounds like a great idea, let me see if I can leverage this for my calculation, if not I may show you the code no problem. Yes the signal is actioned after -say- nine subsequent closes < than close for days ago, and subsequent conditions comparing following close to previous high/lows...the results of all that will be a candle signalling exhaustion of trend. Its a TD sequential. Thank you for your help, this is my first Python project

**Comment by manuelwithbmw** [comment]
Hi @DanielGoldfarb 
I have the feeling I cannot call make_addplot() more than 10 times? Is there a limitation?
I am facing a lot of the below that prevent me from plotting, not sure this is related to the number of make_addplot(): 

_Traceback (most recent call last):
    mpf.plot(df, type='candle', ylabel='Price', ylabel_lower='Volume',
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/mplfinance/plotting.py", line 35, in decorator
    return func(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/mplfinance/plotting.py", line 336, in plot
    ymhi = math.log(max(math.fabs(max(yd)),1e-7),10)
**ValueError: max() arg is an empty sequence_**

Basically I have something like:

apd = [mpf.make_addplot(sell_signal1, scatter=True,
                        markersize=50, marker='$1$', mpf.make_addplot(sell_signal2, scatter=True,
                        markersize=50, marker='$2$', mpf.make_addplot(sell_signal3, scatter=True,
                        markersize=50, marker='$3$', ... , mpf.make_addplot(sell_signal9, scatter=True, ...
                        markersize=50, marker='$9$', .. .)]

but if I exceeds this list I have many **ValueError: max() arg is an empty sequence_** ... or these errors refer to a different problem in my code and I am not seeing the root of it. I keep trying

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw ,
 I am inclined to think this has nothing to do with the _number_ of addplots in your list.  Certainly there is nothing intentional in the code to limit the number.   

The exception is being raised in a part of the code that uses logarithms to determine the "order of magnitude" of the addplot data, to decide whether or not it is the same order of magnitude as the original OHLC data.  If it is not the same order of magnitude, then it will put the addplot data on a "secondary_y" axis.  The log is determined for both the min and max of each set of addplot data.

The exception is complaining that the `max` function is being passed an empty list (sequence) of data.  Just before the line of code that takes the `max`, there is a line of code that generates this list by stripping NaNs out of your addplot data: `yd = [y for y in ydata if not math.isnan(y)]` .   

It seems to me, at a glance, that the only way for that list to be empty is **if, for one of your addplot data sets, you passed in a sequence where the _entire sequence_ was NaN.**

**Please check to see if this may be the case for any of your addplot data sets.**

If it is not the case, **then I may have a bug in the algorithm.**  If so, then if you can provide both your data and your code, I will attempt to debug it.

All the best.  --Daniel

**Comment by manuelwithbmw** [comment]
Sure, let me have a look at these exceptions more deeply, it could well be a full list of NaNs. Also, it seems to me exceptions are df dependent (so yes it may depend on particular values, prices, calculations for that particular time series, while ok for others). In my code, I have a lot of placeholders to check the logic and the steps of the underlying algorithm so I am quite confident the prob is on the plotting side. Will get back later today! Thank you, very useful!

**Comment by DanielGoldfarb** [comment]
A couple points that _may be_ relevant to this issue (once we figure out the cause for the exception):

- Just got a request to **[ignore empty addplots](https://github.com/matplotlib/mplfinance/issues/103)**, the idea being that for users who are generating signals, it keeps their code simpler if their signal generator simply returns empty (when their are no signals) and they don't have to _check_ for empty (rather they just continue appending addplots to their list of addplots).  This may be relevent if indeed you are getting a list full of NaNs.

- We may be able to **[use this to allow passing a list of markers](https://github.com/matplotlib/matplotlib/issues/11155#issuecomment-385939618)** (provided the list of markers is the exact same length as the sequence of data to plot).  This may allow you make fewer addplot calls.

**Comment by manuelwithbmw** [comment]
- Interesting, I read the 'ignore empty' one. Yes from a first check 2 of the lists of data to plot where all NaNs (no signal triggered for that day), the other lists correctly had just one value (1 )(signal triggered for that day), so it seems it depended on this. Will check more instances later.
- Yes a list of markers in may case would do the trick I suppose. Number of markers are exactly =  number of data to plot, so yes same length. Basically I am using the marker as single specific signal and I need to plot several of them

**Comment by manuelwithbmw** [comment]
@DanielGoldfarb 
Maybe not the purest solution but for now, just forcing first element to be zero in those lists, I can plot OK all addplots (you just have not to look at that first candle when all markers overlap :)) and I am happy with that. While maybe waiting for a cleaner solution

signal_1 = np.full(len(signal1), np.nan)
signal_1[0] = 1

**Comment by fxhuhn** [comment]
@manuelwithbmw 
> Hope it helps @fxhuhn !

Thanks, I'm still trying to add it to my charts.

**Comment by fxhuhn** [comment]
@manuelwithbmw 
> * Interesting, I read the 'ignore empty' one. Yes from a first check 2 of the lists of data to plot where all NaNs (no signal triggered for that day), the other lists correctly had just one value (1 )(signal triggered for that day), so it seems it depended on this. Will check more instances later.

It seems we have a similar approach. I'm glad to see my idea helped you with your problem-solving.

**Comment by fxhuhn** [comment]
![ads_de](https://user-images.githubusercontent.com/49436733/79945165-84a61500-846d-11ea-8d7c-ec3ac876bdd5.png)
I'm playing with different colors and triangles. Besides the numbers, the idea with the arrows is very interesting for me.

**Comment by manuelwithbmw** [comment]
@DanielGoldfarb sorry to be a pain. All is good now, every charts plots OK except in one single case where it depends on the timeframe i chose, I almost often have the below messages and series are not all NaNs anymore. Is there any other thing I can check on the max() function functionality? Thanks

Again on:

mpf.plot(df, type='candle', ylabel='Price', ylabel_lower='Volume',
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/mplfinance/plotting.py", line 35, in decorator
return func(*args, **kwargs)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/mplfinance/plotting.py", line 336, in plot
ymhi = math.log(max(math.fabs(max(yd)),1e-7),10)
ValueError: max() arg is an empty sequence

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw ,
I'm not clear on what you are asking.   Unless and until we decide to modify the `mpf.plot()` code to accept an `addplot` argument that contains no data (or all NaNs) then you need to make sure you are not passing in a list or Series, etc. of all NaN's as the first argument to `make_addplot()`

You can test this before calling make addplot, as follows:

Suppose `ydata` is you signal `list` or `Series` (not a dataframe) and you are about to call `mpf.make_addplot(ydata,**kwargs)`, then do something like this:

    if 0 < len( [y for y in ydata if not math.isnan(y)] ):
        apd = mpf.make_addplot(ydata,...)
    else:
        pass    # or, alternatively, print a warning here that you have all NaNs

(If `ydata` is a dataframe, then you just have to do the `0 < len([y for y in ydata if not math.isnan(y)])` check for every column in the dataframe.)

I hope that helps.  Please let me know if this is not clear.

**Comment by manuelwithbmw** [comment]
@DanielGoldfarb sorry Daniel, I just realised simply I was having those errors, because for specific timeframes, I had no signals at all generated by my algorithm (hence all Series -yes they are not dataframes I'm my code- were NaNs. I wasn't expecting there could be such cases in my trading spectrum. I just need to expand the time horizon, the scope of calculations wide enough and the plots are OK.  Apologies if you have wasted time

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw  no problem;  as suggested above, you may want to put a check in your code just before calling `make_addplot()`.  even if all time horizons are good now, it may change in the future with different data sets and/or if you modify the code someday.

**Comment by manuelwithbmw** [comment]
I have added your check code! Thank you

**Comment by char101** [comment]
An example on how to add percentage on top/bottom of the candlebars

```python
def _add_candlestick_labels(ax, ohlc):
    transform = ax.transData.inverted()
    # show the text 10 pixels above/below the bar
    text_pad = transform.transform((0, 10))[1] - transform.transform((0, 0))[1]
    percentages = 100. * (ohlc.close - ohlc.open) / ohlc.open
    kwargs = dict(horizontalalignment='center', color='#000000')
    for i, (idx, val) in enumerate(percentages.items()):
        if val != np.nan:
            row = ohlc.loc[idx]
            open = row.open
            close = row.close
            if open < close:
                ax.text(i, row.high + text_pad, np.round(val, 1), verticalalignment='bottom', **kwargs)
            elif open > close:
                ax.text(i, row.low - text_pad, np.round(val, 1), verticalalignment='top', **kwargs)

figure, axlist = mpf.plot(ohlc, ..., show_nontrading=False, returnfig=True)
_add_candlestick_labels(axlist[0], ohlc)
```

**Comment by DanielGoldfarb** [comment]
![image](https://user-images.githubusercontent.com/11164790/82761057-b04a5100-9dc5-11ea-93e2-dfba0ef00095.png)

---

@char101 

Charles, once again, very creative.  Thanks.
I may integrate this into mplfinance.  Much appreciated.  --Daniel

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw 

`mpf.make_addplot()` now supports a sequence of markers.  **[See here for some examples](https://nbviewer.jupyter.org/github/matplotlib/mplfinance/blob/master/examples/scratch_pad/scatter_multiple_markers.ipynb)**.
Please try it out and let me know if it works for you.

**Comment by manuelwithbmw** [comment]
Hi Daniel,
this looks amazing thank you. You got such a good memory of everything.

PS: I had in mind to write you a generic question:
I am following all your library enhancements with functions and kwargs etc. I am also updating mplfinance to your last release to get them working. In general, I still struggle to understand what is the best way to get all information about new 'functions', new active **kwargs, how to use them, where I can derive their main logic? I understand this is constantly ongoing/changing so examples in notebooks are the last thing you can think of.
However what is the best to get new features' info, is there any /help, original code, where I can nosedive? Thank you for everything you are making this really great

Manuel

**Comment by char101** [comment]
Mplfinance code is actually quite compact, you can look at the code directly, at least that is what I have done. It might be a little difficult if you are not used to matplotlib API but it will get easier the more you research into it.

Also a suggestion to @DanielGoldfarb , if you could format your code with [black](https://github.com/psf/black) or [yapf](https://github.com/google/yapf), it will help other people navigate the code.

**Comment by manuelwithbmw** [comment]
> Mplfinance code is actually quite compact, you can look at the code directly, at least that is what I have done. It might be a little difficult if you are not used to matplotlib API but it will get easier the more you research into it.
> 
> Also a suggestion to @DanielGoldfarb , if you could format your code with [black](https://github.com/psf/black) or [yapf](https://github.com/google/yapf), it will help other people navigate the code.

Ok so I just go into Code tab and looks for the latest commits. Thanks

**Comment by DanielGoldfarb** [comment]
@manuelwithbmw as [Charles](@char101) mentioned, looking at the code can help.   Looking at latest commits may help, _but more importantly I recommend_ looking at the various internal "valid kwargs" methods (for example, [_valid_plot_kwargs()](https://github.com/matplotlib/mplfinance/blob/master/src/mplfinance/plotting.py#L70-L230)).

I also regularly post a brief summary of updates in the [**RELEASE NOTES**](https://github.com/matplotlib/mplfinance/blob/master/RELEASE_NOTES.md).

Finally, almost everything mentioned in the [**RELEASE NOTES**](https://github.com/matplotlib/mplfinance/blob/master/RELEASE_NOTES.md) will have a specific example in one of the [**example notebooks in the Tutorials section**](https://github.com/matplotlib/mplfinance#contents-and-tutorials) (which can also be [found here](https://github.com/matplotlib/mplfinance/tree/master/examples))

**Comment by DanielGoldfarb** [comment]
@char101 
> ... if you could format your code with black or yapf ...

Thanks for the idea.  I will look into it.

**Comment by luongjames8** [comment]
Hi @manuelwithbmw, I am looking to implement the basic DeMark indicators. Your first chart looks like DeMark, but from the thread I gather you're doing some custom with it instead. Just wondering, have you implemented the basic DeMark indicator? If so, is that something you're willing to share? Otherwise, it is my next project. hah.

@DanielGoldfarb - Do you have any plans for a modular way of adding indicators to mplfinance? I would imagine that there is some overlap in terms of the technical studies people are working on implementing?

**Comment by DanielGoldfarb** [comment]
@luongjames8  -- I definitely have in mind to provide a modular or easy way to use third-party and/or user-provided indicators (that can be shared among various users) within mplfinance.  

I have not yet given a lot of thought to how that would be implemented, because presently there remains a few more basic features (labels and annotations, for example) that probably should be implemented first.  (In fact, another contributor is working on labels as we speak).

Here are my "not much yet" thoughts  on the implementation:  I am very reluctant to add any dependencies on other libraries/packages, unless it can be done in a way that is *optional dependency*; perhaps a kwarg to implement the indicator, which first checks for the installing of the required package and raises an exception if not installed.  Alternatively we could just provide a simpler interface for passing in indicators, as either an indicator function, or as data (i.e. an interface somehow simpler than the existing `make_addplot()` interface; one that is specifically designed for indicators - I have in mind one possibly way to do this, where the user passes in a series of *buy and sell signals*, where the *value* of the signal indicates how much to buy (positive values) or sell (negative values)).  This latter approach is the way I am currently leaning.

If you have any thoughts as to how you imagine it to be implemented -- especially in terms of the user interface, or calling convention, for "a modular way of adding indicators to mplfinance" -- please feel free to express your thoughts here.  I am open to suggestions.

**Comment by luongjames8** [comment]
Please give me some time to think about this.

**Comment by orangeasterisk** [comment]
> An example on how to add percentage on top/bottom of the candlebars
> 
> ```python
> def _add_candlestick_labels(ax, ohlc):
>     transform = ax.transData.inverted()
>     # show the text 10 pixels above/below the bar
>     text_pad = transform.transform((0, 10))[1] - transform.transform((0, 0))[1]
>     percentages = 100. * (ohlc.close - ohlc.open) / ohlc.open
>     kwargs = dict(horizontalalignment='center', color='#000000')
>     for i, (idx, val) in enumerate(percentages.items()):
>         if val != np.nan:
>             row = ohlc.loc[idx]
>             open = row.open
>             close = row.close
>             if open < close:
>                 ax.text(i, row.high + text_pad, np.round(val, 1), verticalalignment='bottom', **kwargs)
>             elif open > close:
>                 ax.text(i, row.low - text_pad, np.round(val, 1), verticalalignment='top', **kwargs)
> 
> figure, axlist = mpf.plot(ohlc, ..., show_nontrading=False, returnfig=True)
> _add_candlestick_labels(axlist[0], ohlc)
> ```



> An example on how to add percentage on top/bottom of the candlebars
> 
> ```python
> def _add_candlestick_labels(ax, ohlc):
>     transform = ax.transData.inverted()
>     # show the text 10 pixels above/below the bar
>     text_pad = transform.transform((0, 10))[1] - transform.transform((0, 0))[1]
>     percentages = 100. * (ohlc.close - ohlc.open) / ohlc.open
>     kwargs = dict(horizontalalignment='center', color='#000000')
>     for i, (idx, val) in enumerate(percentages.items()):
>         if val != np.nan:
>             row = ohlc.loc[idx]
>             open = row.open
>             close = row.close
>             if open < close:
>                 ax.text(i, row.high + text_pad, np.round(val, 1), verticalalignment='bottom', **kwargs)
>             elif open > close:
>                 ax.text(i, row.low - text_pad, np.round(val, 1), verticalalignment='top', **kwargs)
> 
> figure, axlist = mpf.plot(ohlc, ..., show_nontrading=False, returnfig=True)
> _add_candlestick_labels(axlist[0], ohlc)
> ```

Hi @char101, 
I'm trying to implement your code, but using dataframes retrieved from the yfinance module. But receive this error: 
_ValueError: Image size of 59405596x487 pixels is too large. It must be less than 2^16 in each direction._

I've realised this error is related to the ax.text() line. And managed to pass the error by adding the parameter "transform = ax.transAxes" to it, but the outputted plot has the percentages way up and outside of the plot https://imgur.com/a/lDUXd9Q.

I was wondering if you could help me out with trying implement your code?

`def _add_candlestick_labels(ax, ohlc):
    transform = ax.transData.inverted()
    # show the text 10 pixels above/below the bar
    text_pad = transform.transform((0, 10))[1] - transform.transform((0, 0))[1]
    percentages = 100. * (ohlc['Close'] - ohlc['Open']) / ohlc['Open']
    print(percentages)
    kwargs = dict(horizontalalignment='center', color='#000000')
    for i, (idx, val) in enumerate(percentages.items()):
        if val != np.nan:
            row = ohlc.loc[idx]
            #print(row)
            open = row.Open
            close = row.Close
            if open < close:
                ax.text(i, row.High+text_pad, np.round(val, 1), verticalalignment='bottom', **kwargs)
                
            elif open > close:
                ax.text(i, row.Low-text_pad, np.round(val, 1), verticalalignment='top', **kwargs)
                
`

**Comment by char101** [comment]
AFAIK only [`set_size_inches`](https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure.set_size_inches) has anything to do with the image size. That code above no matter how big the value you pass should not effect the image size. The default DPI is 100, so you'll just need to divide the output image size in pixels by 100 to the the inces.

**Comment by manuelwithbmw** [comment]
> Hi @manuelwithbmw, I am looking to implement the basic DeMark indicators. Your first chart looks like DeMark, but from the thread I gather you're doing some custom with it instead. Just wondering, have you implemented the basic DeMark indicator? If so, is that something you're willing to share? Otherwise, it is my next project. hah.

Hi @luongjames8 , it is all done, customised and embedded in an whole trad system, sorry cannot share this. Sorry for late reply

**Comment by orangeasterisk** [comment]
> AFAIK only [`set_size_inches`](https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure.set_size_inches) has anything to do with the image size. That code above no matter how big the value you pass should not effect the image size. The default DPI is 100, so you'll just need to divide the output image size in pixels by 100 to get the inches.

Many thanks for the swift reply @char101. Are you saying I have to edit the image size that is outputted from "mpf.show()"?
I've added code below that demonstrates how I'm trying to implement your code. 


```
data = yf.download(tickers="FB")

fig = mpf.figure(figsize=(1,1)) 
ax1 = fig.add_subplot(2,1,1) 
ax2 = fig.add_subplot(2,1,2)

mpf.plot(data, 
         ax=ax1, 
         volume=ax2
         )

_add_candlestick_labels(ax1,  data)

mpf.show()
```

**Comment by char101** [comment]
> > AFAIK only [`set_size_inches`](https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure.set_size_inches) has anything to do with the image size. That code above no matter how big the value you pass should not effect the image size. The default DPI is 100, so you'll just need to divide the output image size in pixels by 100 to get the inches.
> 
> Many thanks for the swift reply @char101. Are you saying I have to edit the image size that is outputted from "mpf.show()"?
> I've added code below that demonstrates how I'm trying to implement your code.
> 
> ```
> data = yf.download(tickers="FB")
> 
> fig = mpf.figure(figsize=(1,1)) 
> ax1 = fig.add_subplot(2,1,1) 
> ax2 = fig.add_subplot(2,1,2)
> 
> mpf.plot(data, 
>          ax=ax1, 
>          volume=ax2
>          )
> 
> _add_candlestick_labels(ax1,  data)
> 
> mpf.show()
> ```

Maybe you could try adding `fig.set_sizes_inches(3, 4)` before `mpf.show()`.

**Comment by char101** [comment]
> > Hi @manuelwithbmw, I am looking to implement the basic DeMark indicators. Your first chart looks like DeMark, but from the thread I gather you're doing some custom with it instead. Just wondering, have you implemented the basic DeMark indicator? If so, is that something you're willing to share? Otherwise, it is my next project. hah.
> 
> Hi @luongjames8 , it is all done, customised and embedded in an whole trad system, sorry cannot share this. Sorry for late reply

IMO you can use use a [scatter ](https://matplotlib.org/3.3.3/api/_as_gen/matplotlib.axes.Axes.scatter.html) plot while adjusting the y points to display those markers.
